descriptors/web_app.js

```diff
@@ -0,0 +1,202 @@
+            descriptor.uuid);
+          assertEquals(characteristic, descriptor.characteristic,
+            'descriptor characteristic mismatch');
+          assertEquals(null, descriptor.value, 'value should be null');
+          const value = await descriptor.readValue();
+          const decoder = new TextDecoder('utf-8');
+          let strVal = decoder.decode(value);
+          assertEquals('Test UINT32 characteristic', strVal);
+
+          strVal = decoder.decode(descriptor.value);
+          assertEquals('Test UINT32 characteristic', strVal);
+
+          const encoder = new TextEncoder('utf-8');
+          try {
+            await descriptor.writeValue(encoder.encode('Something else'));
+          } catch (e) {
+            assertEquals('NotSupportedError', e.name);
+          }
+          break;
+        case characteristicWithoutDescriptorUUID:
+          try {
+            await characteristic.getDescriptors();
+            logError(`Characteristic ${characteristic.uuid} should have no descriptors`);
+          } catch (e) {
+            assertEquals('NotFoundError', e.name);
+          }
+          break;
+        default:
+          logError(`Unexpected characteristic: ${characteristic.uuid}`);
+      }
+    }
+  }
+
+  // Verify descriptors for Nordic UART Service (NUS)
+  // https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/include/bluetooth/services/nus.html
+  const verifyNordicServiceDescriptors = async (service) => {
+    assertEquals(nordicUARTServiceUUID, service.uuid, 'incorrect service UUID');
+
+    const characteristics = await service.getCharacteristics();
+    assertEquals(2, characteristics.length, 'expected 2 characteristic');
+    for (characteristic of characteristics) {
+      switch (characteristic.uuid) {
+        case '6e400002-b5a3-f393-e0a9-e50e24dcca9e': // RX
+          try {
+            await characteristic.getDescriptors();
+            throw 'RX Characteristic should have no descriptors';
+          } catch (e) {
+            assertEquals('NotFoundError', e.name);
+          }
+          break;
+        case '6e400003-b5a3-f393-e0a9-e50e24dcca9e': // TX
+          const descriptors = await characteristic.getDescriptors();
+          assertEquals(1, descriptors.length);
+          const descriptor = descriptors[0];
+          assertEquals(
+            BluetoothUUID.getDescriptor('gatt.client_characteristic_configuration'),
+            descriptor.uuid);
+          assertEquals(characteristic, descriptor.characteristic,
+            'descriptor characteristic mismatch');
+
+          // Not currently subscribed to notifications, so expect zero
+          // subscribers to the RX characteristic.
+          let dataView = await descriptor.readValue();
+          let value = new Uint8Array(dataView.buffer);
+          assertEquals(2, value.length);
+          assertEquals(0, value[0]);
+          assertEquals(0, value[1]);
```

Comment:
Since the CCCD is a bitfield I would read this as a single little-endian 16-bit value,


```js
assertEquals(2, dataView.byteLength);
assertEquals(0x0000, dataView.getUint16(0, true));
```